disallowKeywordsInComments: disallow todo and fixme

joesepi opened the pull request:
@mrjoelkemp

joesepi commented:
I did a couple `commit --amend`s after an initial style check fail. That's part of the reason why you see two coverall reports.

joesepi commented:
This fix is for this feature request:
https://github.com/jscs-dev/node-jscs/issues/646

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,69 @@
+ */
+
+var assert = require('assert');
+
```
Name this variable `keywordRegex`. More descriptive.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,69 @@
+
+                    // adjust line location for reporter
+                    // comment is split by new line and iterated over
+                    errorLoc.line = comment.loc.start.line + index;
```
Inline this object directly into the error.add call. No need to define as empty object first, nor even assign to a var.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,69 @@
+    configure: function(disallowKeywordsInComments) {
+        assert(
+            disallowKeywordsInComments === true,
+            'disallowKeywordsInComments option disallows certain keywords in your comments'
```
This error message should say what the required value is. See other rules for suggestions.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,69 @@
+            var splitComment = comment.value.split('\n');
+
+            splitComment.forEach(function(commentNode, index) {
+                var matchIndex = commentNode.search(r);
```
What if the word TODO appears twice in the same line? You'll only report one error...

Which I think is fine, but it highlights the fact that getting the exact location of the offending word is probably not important. You could simplify this rule greatly if you just reported the start of the comment as the error location.

mikesherov commented:
@joesepi, thanks for contributing, friend!

Good work. Found a few nits to address, but also:

 * commit message should say which issue it closes: 

```
disallowKeywordsInComments: new rule 

Fixes #646
```

 * the rule should take in `true` or a string that represents a regex to use instead of the default `"/Joel/"`.

Thanks again!

joesepi commented:
@mrjoelkemp and @mikesherov and whoever else is interested. I updated the rule and test. I moved the bit where the comment is broken up by lines (for better carat placement) out of the check itself. My intent, after discussing it with Joel earlier, is to move it to a comment util or somewhere similar. I thought this would be a good point to get feedback and if things are good, I can pull it out entirely. Thanks!!

mikesherov commented:
@joesepi you have a few style violations in your code, which is why travis is failing. Running `npm test` locally will reveal the erros so you don't have to wait for travis.

mrjoelkemp commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,131 @@
+    if (!suppliedPattern) {
+        pattern = '\\b(';
+        pattern += keywords.join('|');
+        pattern += ')\\b';
```
Why not just munge all together on one line? 

I'll take care of that post-merge.

mrjoelkemp commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,131 @@
+    configure: function(disallowKeywordsInComments) {
+        // use default keywords
+        if (disallowKeywordsInComments === true) {
+            message = buildMessage();
```
We typically store data on the rule instance to avoid shared state across instances: `this._message = buildMessage()`.

mrjoelkemp commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,131 @@
+        var lastIndex = 0;
+
+        if (matches) {
+            matches.map(function(match) {
```
Alternative
```js
errorArray = errorArray.concat(matches.map(function(match) {
   ... 
   return {
     line: ...,
     column: ...
   };
}))
```

mrjoelkemp commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,131 @@
+
+var getCommentErrors = function(comment) {
+    var splitComment = comment.value.split('\n');
+    var errorArray = [];
```
`errors` implies it's an array

mrjoelkemp commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,131 @@
+        var matches = commentNode.match(keywordRegEx);
+        var lastIndex = 0;
+
+        if (matches) {
```
`if (!matches) { return; }`

zxqfox replied:
I want this as rule! ;-D

mrjoelkemp commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,131 @@
+            var errorCheck = comment.value.search(keywordRegEx);
+
+            if (errorCheck >= 0) {
+                var errorArray = getCommentErrors(comment);
```
Consider removing the temp var:

```js
getCommentErrors(comment)
.forEach(...)
```

mrjoelkemp commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,131 @@
+        file.getComments().forEach(function(comment) {
+            var errorCheck = comment.value.search(keywordRegEx);
+
+            if (errorCheck >= 0) {
```
Consider introducing a better explaining variable like: `hasViolations = comment.value.search(keywordRegEx) !== -1`

joesepi commented:
@mikesherov and @mrjoelkemp and others: all good comments. Thank you. I have updated the code and ran npm test locally so Travis shouldn't complain. Please recheck and comment if need be.

zxqfox commented on lib/rules/disallow-keywords-in-comments.js:
````diff
@@ -0,0 +1,119 @@
+ *
+ * ```js
+ * "disallowKeywordsInComments": true
+ * "disallowKeywordsInComments": '\b(word1|word2)\b'
````
I'm not sure, but feels like it should be `double-slash b` (`"\\b(...)\\b"`) and in double-quotes ('cause of JSON). Right?

mikesherov replied:
yes. that's right.

zxqfox commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+
+module.exports.prototype = {
+    configure: function(disallowKeywordsInComments) {
+        if (disallowKeywordsInComments === true) {
```
Suggest switch-case here:
```js
switch (true) {
  case disallowKeywordsInComments === true:
  ...
  default:
    assert(false, ...
}
```

mikesherov replied:
especially considering leveraging the fall through:

```
witch (true) {
  case Array.isArray(disallowKeywordsInComments):
    keywords = disallowKeywordsInComments;
    /*falls through*/
  case disallowKeywordsInComments === true:
    this._message = buildMessage();
    keywordRegEx = buildKeywordRegEx();
    break;
  ...
  default:
    assert(false, ...
}

zxqfox commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+var buildKeywordRegEx = function(suppliedPattern) {
+    var pattern;
+
+    if (!suppliedPattern) {
```
mmm... maybe:
```js
pattern = suppliedPattern || ('\\b(' + keywords.join('|') + ')\\b');

mikesherov replied:
agree with @zxqfox

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+var keywords = ['todo', 'fixme'];
+
+var keywordRegEx;
+var buildKeywordRegEx = function(suppliedPattern) {
```
why var not a function declaration?

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+    return new RegExp(pattern, 'gi');
+};
+
+var buildMessage = function() {
```
why not function declaration?

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+    return 'Comments cannot contain the following keywords: ' + keywordsList;
+};
+
+var getCommentErrors = function(comment) {
```
why not function declaration?

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+// default keywords
+var keywords = ['todo', 'fixme'];
+
+var keywordRegEx;
```
this should a member property, not a global variable.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+        file.getComments().forEach(function(comment) {
+            if (comment.value.search(keywordRegEx) === -1) { return; }
+            getCommentErrors(comment).forEach(function(errorObj) {
+                errors.add(this._message, errorObj);
```
`this` is the wrong thing here... need to bind up.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+            if (comment.value.search(keywordRegEx) === -1) { return; }
+            getCommentErrors(comment).forEach(function(errorObj) {
+                errors.add(this._message, errorObj);
+            });
```
`}, this);`

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,119 @@
+            getCommentErrors(comment).forEach(function(errorObj) {
+                errors.add(this._message, errorObj);
+            });
+        });
```
`}, this);`

joesepi commented:
You guys, these are all great comments. Thank you! The code has improved greatly. Check it out!

@mrjoelkemp @mikesherov @zxqfox

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,103 @@
+
+    check: function(file, errors) {
+        file.getComments().forEach(function(comment) {
+            if (comment.value.search(this._keywordRegEx) === -1) { return; }
```
You should do this check inside of the `getCommentErrors` function so your not `search`ing and then `matching`. Just `match` and there are zero matches, return empty array.

mikesherov commented on test/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,49 @@
+    beforeEach(function() {
+        checker = new Checker();
+        checker.registerDefaultRules();
+        checker.configure({ disallowKeywordsInComments: true });
```
Do one `describe` per option value. Right now you are only testing `true`. Likely why coverage is listed as going down.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,103 @@
+
+var assert = require('assert');
+
+var getCommentErrors = function(comment, keywordRegEx) {
```
function declaration instead of assigning to var here please.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,103 @@
+    var errors = [];
+
+    splitComment.forEach(function(commentNode, index) {
+        var lineIndex = index;
```
Why create this variable? Just name index lineIndex in the function params.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,103 @@
+        errors = errors.concat(matches.map(function(match) {
+            var foundAtIndex;
+
+            lastIndex = lastIndex === 0 ? lastIndex : lastIndex + 1;
```
If you initialize `lastIndex` to `-1` instead of `0`, this whole line is just `lastIndex++`.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,103 @@
+            var foundAtIndex;
+
+            lastIndex = lastIndex === 0 ? lastIndex : lastIndex + 1;
+            foundAtIndex = commentNode.indexOf(match, lastIndex);
```
No need for `foundAtIndex`, just assign directly to `lastIndex` and use it everywhere you used `foundAtIndex` below.

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,103 @@
+            //   +2 finds accurate carat position on opening line comment
+            return {
+                line: comment.loc.start.line + lineIndex,
+                column: lineIndex <= 0 ? foundAtIndex += 2 : foundAtIndex
```
`<=` -> `===`

mikesherov commented on lib/rules/disallow-keywords-in-comments.js:
```diff
@@ -0,0 +1,103 @@
+                // use the array of strings provided to build RegExp pattern
+                this._keywords = disallowKeywordsInComments;
+                /* falls through */
+            case disallowKeywordsInComments:
```
Make sure to test the string option, because I think it'll never get hit.

mikesherov commented:
*ping* @joesepi

joesepi commented:
Sorry folks. I wanted to wrap this up before I went on vacation but it didn't happen. I will try to finish it in the coming days.

mikesherov commented:
Landed with the fixes mentioned! Thanks again for the hard work here @joesepi !
